specs/core/1_beacon-chain.md

```diff
@@ -0,0 +1,736 @@
+```
+
+#### `get_indexed_attestation`
+
+```python
+def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation) -> AttestationAndCommittee:
+    committee = get_beacon_committee(beacon_state, attestation.data.slot, attestation.data.index)
+    return AttestationAndCommittee(committee, attestation)
+```
+
+#### `get_updated_gasprice`
+
+```python
+def get_updated_gasprice(prev_gasprice: Gwei, length: uint8) -> Gwei:
+    if length > TARGET_SHARD_BLOCK_SIZE:
+        delta = (prev_gasprice * (length - TARGET_SHARD_BLOCK_SIZE) 
+                 // TARGET_SHARD_BLOCK_SIZE // GASPRICE_ADJUSTMENT_COEFFICIENT)
+        return min(prev_gasprice + delta, MAX_GASPRICE)
+    else:
+        delta = (prev_gasprice * (TARGET_SHARD_BLOCK_SIZE - length)
+                 // TARGET_SHARD_BLOCK_SIZE // GASPRICE_ADJUSTMENT_COEFFICIENT)
+        return max(prev_gasprice, MIN_GASPRICE + delta) - delta
+```
+
+#### `get_start_shard`
+
+```python
+def get_start_shard(state: BeaconState, slot: Slot) -> Shard:
+    # TODO: implement start shard logic
+    return Shard(0)
+```
+
+#### `get_shard`
+
+```python
+def get_shard(state: BeaconState, attestation: Attestation) -> Shard:
+    return Shard((attestation.data.index + get_start_shard(state, attestation.data.slot)) % ACTIVE_SHARDS)
+```
+
+#### `get_offset_slots`
+
+```python
+def get_offset_slots(state: BeaconState, start_slot: Slot) -> Sequence[Slot]:
+    return [Slot(start_slot + x) for x in SHARD_BLOCK_OFFSETS if start_slot + x < state.slot]
+```
+
+
+### Predicates
+
+#### Updated `is_valid_indexed_attestation`
+
+Note that this replaces the Phase 0 `is_valid_indexed_attestation`.
+
+```python
+def is_valid_indexed_attestation(state: BeaconState, indexed_attestation: AttestationAndCommittee) -> bool:
+    """
+    Check if ``indexed_attestation`` has valid indices and signature.
+    """
+
+    # Verify aggregate signature
+    all_pubkeys = []
+    all_message_hashes = []
+    aggregation_bits = indexed_attestation.attestation.aggregation_bits
+    assert len(aggregation_bits) == len(indexed_attestation.committee)
+    for i, custody_bits in enumerate(indexed_attestation.attestation.custody_bits):
+        assert len(custody_bits) == len(indexed_attestation.committee)
+        for participant, abit, cbit in zip(indexed_attestation.committee, aggregation_bits, custody_bits):
+            if abit:
+                all_pubkeys.append(state.validators[participant].pubkey)
+                # Note: only 2N distinct message hashes
+                all_message_hashes.append(hash_tree_root(
+                    AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, cbit)
+                ))
+            else:
+                assert not cbit
+        
+    return bls_verify_multiple(
+        pubkeys=all_pubkeys,
+        message_hashes=all_message_hashes,
+        signature=indexed_attestation.signature,
```

Comment:
Same comment as the previous one `indexed_attestation.attestation.signature` is probably the correct implementation since `indexed_attestation` is encapsulating the `attestation`

Reply:
thanks!